### PR TITLE
Block Bindings: Create utils to update or remove bindings

### PR DESCRIPTION
**Pull request by SantosGuillamot**
## What?
This pull requests explores the idea of exposing two new block bindings utils to easily add/remove/update bindings in the editor:
* updateBlockBindings: It tries to follow the same approach as `updateBlockAttributes`. You pass an object with all the bindings to update, and to remove one users can pass `undefined`.
* removeAllBlockBindings: It just removes all the existing bindings of a block.

## Why?
The same code was used for the custom fields UI and pattern overrides. Additionally, it could be useful for external developers in case they want to create their own UIs.

## How?
Creating a `useBlockBindingsUtils` hook that return the needed function. It is inspired by other hooks like [useGradient](https://github.com/WordPress/gutenberg/blob/9373c1447da400019f3a6494029a9a1b60faa4f6/packages/block-editor/src/components/gradients/use-gradient.js#L57).

## Testing Instructions
It doesn't add new functionalities. Automated tests should pass.

### CODE REVIEW

File: packages/block-editor/src/hooks/block-bindings.js [hunk 1]
```diff
@@ -286,7 +231,7 @@ export const BlockBindingsPanel = ( { name, metadata } ) => {
 				} );
 				// Only add source if the list is not empty.
 				if ( sourceList ) {
-					fieldsList[ label ] = { ...sourceList };
+					fieldsList[ sourceName ] = { ...sourceList };
```
Comment (artemiomorales) [inline packages/block-editor/src/hooks/block-bindings.js:234]: Why were we using the `label` here before? It seems like that was brittle. It appears that this change is unrelated to the adding of the utils in this PR, and is just an additional code cleanup, and it looks good. I'm just curious.

**Review by artemiomorales — APPROVED** [review]
Looks good to me!

File: packages/block-editor/src/hooks/block-bindings.js (hunk 1 above)
Comment (cbravobernal) [reply packages/block-editor/src/hooks/block-bindings.js:234]: It seems that we were just using the values of the `registeredSources` object instead of their keys.

I'm not sure if if sources label were always the same as the registeredSources key, but this approach seems safer, as the label should allow whitespaces or uppercases.

**Review by cbravobernal — APPROVED** [review]
LGTM.

File: packages/block-editor/src/private-apis.js [hunk 2]
```diff
@@ -93,4 +94,5 @@ lock( privateApis, {
 	useBlockDisplayTitle,
 	__unstableBlockStyleVariationOverridesWithConfig,
 	setBackgroundStyleDefaults,
+	useBlockBindingsUtils,
```
Comment (gziolo) [inline packages/block-editor/src/private-apis.js:97]: What's the rationale for exposing `useBlockBindingsUtils` as a private API given that it isn't used outside of the package where it is defined?

File: packages/block-editor/src/utils/block-bindings.js [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+	 * import { useBlockBindingsUtils } from '@wordpress/block-editor'
+	 *
+	 * const { removeAllBlockBindings } = useBlockBindingsUtils();
+	 * removeAllBlockBindings();
+	 * ```
+	 */
+	const removeAllBlockBindings = () => {
+		const { metadata } = getBlockAttributes( clientId );
+		const newMetadata = { ...metadata };
+		delete newMetadata.bindings;
```
Comment (gziolo) [inline packages/block-editor/src/utils/block-bindings.js:88]: That could be a single line:
```js
const { metadata: { bindings, ...metadata } } = getBlockAttributes( clientId );

File: packages/block-editor/src/utils/block-bindings.js [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+/**
+ * Internal dependencies
+ */
+import { store as blockEditorStore } from '../store';
+import { useBlockEditContext } from '../components/block-edit';
+
+export function useBlockBindingsUtils() {
+	const { clientId } = useBlockEditContext();
+	const { updateBlockAttributes } = useDispatch( blockEditorStore );
+	const { getBlockAttributes } = useSelect( blockEditorStore );
```
Comment (gziolo) [inline packages/block-editor/src/utils/block-bindings.js:15]: It would be more optimal to have:
```js
const { getBlockAttributes } = useRegistry().select( blockEditorStore );
```

See more in [this guide](https://developer.wordpress.org/news/2024/03/28/how-to-work-effectively-with-the-useselect-hook/#but-call-them-outside-when-you-re-in-an-event-handler).

File: packages/block-editor/src/utils/block-bindings.js [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+			}
+			newBindings[ attribute ] = binding;
+		} );
+
+		const newMetadata = {
+			...metadata,
+			bindings: newBindings,
+		};
+
+		if ( Object.keys( newMetadata.bindings ).length === 0 ) {
```
Comment (gziolo) [inline packages/block-editor/src/utils/block-bindings.js:62]: Here and in other places it would increase the readability if the helper function was present:

```js
function isObjectEmpty( object ) {
	return ! object || Object.keys( object ).length === 0;
}

// later
if ( isObjectEmpty( newMetadata.bindings ) ) {
```

This could also be further refactored so it doesn't require the step to remove the undefined key:
```js
const newMetadata = {
	...metadata,
	...( ! isObjectEmpty( newBindings ) && {
		bindings: newBindings,
	} ),
};
```

**Comment by gziolo** [comment]
Nice refactoring. It looks like two methods allow to accomplish all necessary operations:

```js
const { updateBlockBindings, removeAllBlockBindings ) = useBlockBindingsUtils();

// Add or update a binding.
updateBlockBindings( { attributeName: { source: 'my/source' } );

// Remove a binding - that could also be included in the inline docs.
updateBlockBindings( { attributeName: undefined );

// Remove all block bindings for a block.
removeAllBlockBindings();
```

When opening this API to the public, it might also be worth considering allowing passing a different `clientId` (replicating multiple ids that `updateBlockAttributes` supports is rather too complex for its narrow usage here) to `useBlockBindingsUtils` in the case developers would like to modify arbitrary blocks:

```js
// Remove all block bindings for another block.
const { removeAllBlockBindings ) = useBlockBindingsUtils( anotherClientId );
removeAllBlockBindings();
```

File: packages/block-editor/src/hooks/block-bindings.js (hunk 1 above)
Comment (SantosGuillamot) [reply packages/block-editor/src/hooks/block-bindings.js:234]: > Why were we using the label here before?

I believe the name wasn't exposed before. But this seemed safer to me as well.

**Comment by colorful-tones** [comment]
@SantosGuillamot @gziolo @cbravobernal 

> Additionally, it could be useful for external developers in case they want to create their own UIs.

Curious, was there some boilerplate code y'all used to test this? I wanted to test it, but did not see anything in the notes of this ticket. 🕵️ 🤔

**Comment by SantosGuillamot** [comment]
When this was first added, it was still a private API used by `core/post-meta` and `core/pattern-overrides`, that have their own tests.

You can find an example of how it could be used in [the readme](https://github.com/WordPress/gutenberg/blob/0e17bd28ac8adf8612b9daa2b1d30e1961b7c58d/packages/block-editor/README.md#useblockbindingsutils). However, trying to create a quick code snippet, I've realized that we didn't address passing a `clientId` as suggested [here](https://github.com/WordPress/gutenberg/pull/64102#issuecomment-2306631817). I can work on that in a follow-up PR and share an example of how to test it.

**Comment by SantosGuillamot** [comment]
I've started this pull request to explore that possibility: https://github.com/WordPress/gutenberg/pull/65818

It's hard to test manually because hooks need to run inside components or hooks, but I add some unit tests that should serve as reference and to give more confidence on the APIs.
